gcloud-java-gax/src/main/java/io/gapi/gax/protobuf/PathTemplate.java

```diff
@@ -0,0 +1,880 @@
+            implicitWildcard = true;
+            varName = seg.substring(0, seg.length() - 1).trim();
+            seg = seg.substring(seg.length() - 1).trim();
+          } else {
+            throw new ValidationException("parse error: invalid binding syntax in '%s'", template);
+          }
+        } else {
+          // Looking at something like "{name=wildcard}".
+          varName = seg.substring(0, i).trim();
+          seg = seg.substring(i + 1).trim();
+        }
+        builder.add(Segment.create(SegmentKind.BINDING, varName));
+      }
+
+      // If segment ends with '}', a binding group ends. Remove the brace and remember.
+      boolean bindingEnds = seg.endsWith("}");
+      if (bindingEnds) {
+        seg = seg.substring(0, seg.length() - 1).trim();
+      }
+
+      // Process the segment, after stripping off "{name=.." and "..}".
+      switch (seg) {
+        case "**":
+        case "*":
+          if ("**".equals(seg)) {
+            pathWildCardBound++;
+          }
+          Segment wildcard = seg.length() == 2 ? Segment.PATH_WILDCARD : Segment.WILDCARD;
+          if (varName == null) {
+            // Not in a binding, turn wildcard into implicit binding.
+            // "*" => "{$n=*}"
+            builder.add(Segment.create(SegmentKind.BINDING, "$" + freeWildcardCounter));
+            freeWildcardCounter++;
+            builder.add(wildcard);
+            builder.add(Segment.END_BINDING);
+          } else {
+            builder.add(wildcard);
+          }
+          break;
+        case "":
+          if (!bindingEnds) {
+            throw new ValidationException("parse error: empty segment not allowed in '%s'",
+                template);
+          }
+          // If the wildcard is implicit, seg will be empty. Just continue.
+          break;
+        default:
+          builder.add(Segment.create(SegmentKind.LITERAL, seg));
+      }
+
+      // End a binding.
+      if (bindingEnds) {
+        // Reset varName to null for next binding.
+        varName = null;
+
+        if (implicitWildcard) {
+          // Looking at something like "{var}". Insert an implicit wildcard, as it is the same
+          // as "{var=*}".
+          builder.add(Segment.WILDCARD);
+        }
+        builder.add(Segment.END_BINDING);
+      }
+
+      if (pathWildCardBound > 1) {
+        // Report restriction on number of '**' in the pattern. There can be only one, which
+        // enables non-backtracking based matching.
+        throw new ValidationException(
+            "parse error: pattern must not contain more than one path wildcard ('**') in '%s'",
+            template);
+      }
+
+    }
+
+    if (customVerb != null) {
+      builder.add(Segment.create(SegmentKind.CUSTOM_VERB, customVerb));
+    }
+    return builder.build();
+  }
+
+  // Helpers
+  // =======
+
+  private static String encodeUrl(String text) {
+    try {
+      return URLEncoder.encode(text, "UTF-8");
```

Comment:
Instead of hardcoding the strings you can use [`StandardCharsets.UTF_8.name()`](https://docs.oracle.com/javase/7/docs/api/java/nio/charset/StandardCharsets.html#UTF_8)